### ISSUE TITLE
defaultdict loses its default factory in serializing/deserialising roundtrip

### ISSUE THREAD

**Issue by bluenote10** [pydantic/pydantic#9060]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

It looks like when serializing/deserialising a `BaseModel` that holds a `defaultdict`, its `default_factory` is lost.

This has been previously reported in https://github.com/pydantic/pydantic/issues/4687 for v1 and the special case of a nested defaultdict. In v2 I'm facing the issue even without nesting.

### Example Code

```py
from collections import defaultdict
from pydantic import BaseModel

class MyModel(BaseModel):
    some_dict: defaultdict[str, int] = defaultdict(lambda: 42)

instance_orig = MyModel()
print(instance_orig.some_dict["a"])  # -> 42
print(instance_orig.some_dict["b"])  # -> 42

instance_roundtrip = MyModel.model_validate_json(instance_orig.model_dump_json())
print(instance_roundtrip.some_dict["a"])  # -> 42
print(instance_roundtrip.some_dict["b"])  # -> 42
print(instance_roundtrip.some_dict["c"])  # -> 0
```

The output is:

```
42
42
42
42
0
```

Note that after the serialization/deserialization roundtrip, the specified `default_factory` is lost. This violates the invariant of the class, because it no longer behaves as intended (in this case returning 42 for missing entries).

### Python, Pydantic & OS Version

```Text
pydantic version: 2.6.4
        pydantic-core version: 2.16.3
          pydantic-core build: profile=release pgo=true
                 install path: /home/me/path/to/venv/lib/python3.10/site-packages/pydantic
               python version: 3.10.13 (main, Aug 25 2023, 13:20:03) [GCC 9.4.0]
                     platform: Linux-5.4.0-173-generic-x86_64-with-glibc2.31
             related packages: mypy-1.9.0 typing_extensions-4.10.0
                       commit: unknown
```

**Comment by sydney-runkle** [comment]
@bluenote10,

Huh, certainly seems like a bug. PRs welcome with a fix, and I'll add this to the 2.8 milestone.

**Comment by sydney-runkle** [comment]
Marking as resolved, according to https://github.com/pydantic/pydantic/pull/9543#issuecomment-2155363669.

Probably helpful to read through the whole thought process on the PR to get some more context, but the general idea is that you can use this structure to get the default factory to apply:

```py
some_field: DefaultDict[<key_type>, Annotated[<value_type>, Field(..., default_factory=lambda: <default_value>)]]
```
